ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3ListHandler.java

```
@@ -0,0 +1,151 @@
+   * @param namedBlobListHandler named blob list handler
+   */
+  public S3ListHandler(NamedBlobListHandler namedBlobListHandler) {
+    this.namedBlobListHandler = namedBlobListHandler;
+  }
+
+  /**
+   * Asynchronously get account metadata.
+   * @param restRequest the {@link RestRequest} that contains the request parameters and body.
+   * @param restResponseChannel the {@link RestResponseChannel} where headers should be set.
+   * @param callback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   */
+  public void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      Callback<ReadableStreamChannel> callback) {
+    this.restRequest = restRequest;
+    namedBlobListHandler.handle(restRequest, restResponseChannel, ((result, exception) -> {
+      if (exception != null) {
+        callback.onCompletion(result, exception);
+        return;
+      }
+
+      try {
+        // Convert from json response to S3 xml response as defined in
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_ResponseSyntax.
+        RetainingAsyncWritableChannel writableChannel = new RetainingAsyncWritableChannel();
+        result.readInto(writableChannel, null).get();
+        JSONObject jsonObject = FrontendUtils.readJsonFromChannel(writableChannel);
+        Page<NamedBlobListEntry> page = Page.fromJson(jsonObject, NamedBlobListEntry::new);
+        ReadableStreamChannel readableStreamChannel = serializeAsXml(page);
+        restResponseChannel.setHeader(RestUtils.Headers.DATE, new GregorianCalendar().getTime());
+        restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, "application/xml");
+        restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, readableStreamChannel.getSize());
+        callback.onCompletion(readableStreamChannel, null);
+      } catch (Exception e) {
+        throw new RuntimeException(e);
```

This should be a 503 status. Let's just return a runtime exception here.

---

Done..